url55e48746be_isostorm_RPGame_tgz-pJ8-rpg_MainJ8/src/rpg/creature/Anchor.java

```diff
@@ -0,0 +1,321 @@
+	}
+
+	private final Creature holder;
+	
+	/**
+	 * Returns the holder of this anchor
+	 */
+	@Raw @Basic @Immutable
+	public Creature getHolder() {
+		return holder;
+	}
+	
+	private final String name;
+	
+	/**
+	 * Returns the name of this anchor
+	 */
+	@Immutable @Basic @Raw
+	public String getName() {
+		return name;
+	}
+	
+	private Item item;
+	
+	/**
+	 * Returns the item this anchor holds.
+	 */
+	@Raw @Basic
+	public Item getItem() {
+		return item;
+	}
+	
+	/**
+	 * Sets the item this anchor holds.
+	 * 
+	 * @param  item
+	 *         The new item to be set
+	 * @post   If the given item is an item implementation, 
+	 * 		   the parent of this item is equal to this anchor
+	 * 		   | if( item instanceof ItemImplementation ) then
+	 *         |     item.getParent() == this
+	 * @post   The item of this anchor equals the given item.
+	 *         | new.getItem() == item
+	 * @throws IllegalAddItemException
+	 *         The given item can't be added to this anchor.
+	 *         | !canAddItem(item)
+	 */
+	public void addItem(Item item) throws IllegalAddItemException {
+		if(!canAddItem(item))
+			throw new IllegalAddItemException(this, item);
+		
+		setItem(item);
+		
+		if(item instanceof ItemImplementation)
+			((ItemImplementation)item).setParent(this);
+	}
+	
+	/**
+	 * Check whether the given item can be added to this anchor.
+	 * 
+	 * @return False if the given item is ineffective.
+	 *         | if( item == null) then
+	 *         |    result == false
+	 *         False if the item this anchor contains is effective.
+	 *         | if (getItem() != null) then
+	 *         |    result == false
+	 *         Otherwise false if the given item is an instance of
+	 *         item implementation and has a parent.
+	 *         | if( item instanceof ItemImplementation && ((ItemImplementation)item).hasParent() ) then
+	 *         |    result == false
+	 *         Otherwise the result is equal to whether or not the item can be added to the holder of this anchor.
+	 *         | result == getHolder().canAddItem(item)
+	 *         
+	 */
+	@Override
+	public boolean canAddItem(Item item)
+	{
+		if(item == null)
+			return false;
+		if(getItem() != null)
+			return false;
+		if( item instanceof ItemImplementation && ((ItemImplementation)item).hasParent() )
+			return false;
+		return getHolder().canAddItem(item);
+	}
+	
+	/**
+	 * Swaps the item this anchor holds with the given item.
+	 * 
+	 * @param replacement
+	 *        The item to swap with.
+	 * @throws IllegalAddItemException
+	 *         If the replacement is not effective
+	 *         | replacement == null
+	 * @throws IllegalArgumentException
+	 *         If addItem throws an exception
+	 */
+	public void swap(Item replacement) throws IllegalAddItemException, IllegalArgumentException
+	{
+		if(replacement == null)
+			throw new IllegalArgumentException();
+		Item oldItem = getItem();
+		try
+		{
+			removeItem();
+			addItem(replacement);
+		}
+		catch(IllegalAddItemException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)